provider/provider_test.go

```diff
@@ -87,3 +125,678 @@ func TestProviderCdpRegion(t *testing.T) {
+	}
+}
+
+func TestGetOrDefaultFromEnvWithValue(t *testing.T) {
+	val := types.StringValue("expected")
+	if got := getOrDefaultFromEnv(val, testEnvVariable); got != val.ValueString() {
+		t.Errorf("Assertion faled! Expected: %s, got: %s\n", val.ValueString(), got)
+	}
+}
+
+func TestGetOrDefaultFromEnvWhenValuePresentInEnv(t *testing.T) {
+	val := "expected"
+	t.Setenv(testEnvVariable, val)
+	if got := getOrDefaultFromEnv(types.StringNull(), testEnvVariable); got != val {
+		t.Errorf("Assertion faled! Expected: %s, got: %s\n", val, got)
+	}
+}
+
+func TestGetOrDefaultFromEnvWhenValuePresentInEnvButNotFirst(t *testing.T) {
+	val := "expected"
+	t.Setenv(testEnvVariable, val)
+	if got := getOrDefaultFromEnv(types.StringNull(), "customOtherEnvVariable", testEnvVariable); got != val {
+		t.Errorf("Assertion faled! Expected: %s, got: %s\n", val, got)
+	}
+}
+
+func TestGetOrDefaultBoolFromEnvWithValue(t *testing.T) {
+	val := types.BoolValue(true)
+	if got := getOrDefaultBoolFromEnv(context.Background(), val, testEnvVariable); got != val.ValueBool() {
+		t.Errorf("Assertion faled! Expected: %t, got: %t\n", val.ValueBool(), got)
+	}
+}
+
+func TestGetOrDefaultBoolFromEnvWhenValuePresentInEnv(t *testing.T) {
+	val := "true"
+	t.Setenv(testEnvVariable, val)
+	if got := getOrDefaultBoolFromEnv(context.Background(), types.BoolNull(), testEnvVariable); !got {
+		t.Errorf("Assertion faled! Expected: %t, got: %t\n", true, got)
+	}
+}
+
+func TestGetOrDefaultBoolFromEnvWhenValuePresentInEnvButNotFirst(t *testing.T) {
+	val := "true"
+	t.Setenv(testEnvVariable, val)
+	if got := getOrDefaultBoolFromEnv(context.Background(), types.BoolNull(), "customOtherEnvVariable", testEnvVariable); !got {
+		t.Errorf("Assertion faled! Expected: %t, got: %t\n", true, got)
+	}
+}
+
+func TestResources(t *testing.T) {
+	expectedResources := []func() resource.Resource{
+		environments.NewAwsCredentialResource,
+		environments.NewAwsEnvironmentResource,
+		environments.NewIDBrokerMappingsResource,
+		environments.NewAzureCredentialResource,
+		environments.NewAzureEnvironmentResource,
+		environments.NewGcpEnvironmentResource,
+		environments.NewGcpCredentialResource,
+		datalake.NewAwsDatalakeResource,
+		datalake.NewAzureDatalakeResource,
+		iam.NewGroupResource,
+		datahub.NewAwsDatahubResource,
+		datahub.NewAzureDatahubResource,
+	}
+
+	provider := CdpProvider{testVersion}
+
+	resources := provider.Resources(context.Background())
+
+	if len(resources) > len(expectedResources) {
+		var extra []string
+		for _, f := range resources {
+			has := false
+			for _, expectedResource := range expectedResources {
+				if getPackageAndNameOfResource(f) == getPackageAndNameOfResource(expectedResource) {
+					has = true
+					break
+				}
+			}
+			if !has {
+				extra = append(extra, getPackageAndNameOfResource(f))
+			}
+		}
+		t.Errorf("The following unexpected resource(s) got introduced: %s", strings.Join(extra, ","))
+	} else if len(resources) < len(expectedResources) {
```

Comment:
the two conditions are for making our lives easier to know whether something was added *or* removed unintentionally. there won't be any unnecessary loop iterations. if the 'got' resources is larger than expected then one loop goes, if less, then *another one*. else nothing

Reply:
The problem is that the lenghts of the lists are meaningless to compare since you can have a shorter list and a longer list that still do not contain the elements of each other. For example:
{code}
list1 = {"a", "b", "c"}
list2 = {"a", "b", "d", "e"}
{code}

here, you want to print both "missing c" and additional "d" and "e" in your logic. 

What you need is the Java equivalent of: 

{code}
var additional = Sets.difference(resources, expectedResources);
var missing = Sets.difference(expectedResources, resources);

// print additional
// print missing
{code}

I think the more readable approach would be to write a utility function that takes two []string's (or sets), and returns the difference. Then you can just print the resulting lists like the pseodo-code above.

Reply:
@enis how about this approach?